core/arch/interp.c

```diff
@@ -8007,3 +8252,157 @@ d_r_emulate(dcontext_t *dcontext, app_pc pc, priv_mcontext_t *mc)
+    instr_t *trace_exit_label;
+    app_pc target = 0;
+    app_pc ind_target = 0;
+    reg_id_t scratch;
+    reg_id_t jump_target_reg = DR_REG_NULL;
+    uint indirect_type = 0;
+    int added_size = 0;
+    trace_exit_label = NULL;
+    instr_t *delete_after = NULL;
+    /* We record the original trace end */
+    instr_t *trace_end = instrlist_last(trace);
+
+    LOG(THREAD, LOG_MONITOR, 4, "fixup the indirect trace exit\n");
+
+    for (instr = instrlist_first(trace); instr != trace_end;
+         instr = instr_get_next(instr)) {
+        if (instr_is_exit_cti(instr)) {
+            target = instr_get_branch_target_pc(instr);
+            /* Check for indirect exit. */
+            if (is_indirect_branch_lookup_routine(dcontext, (cache_pc)target)) {
+                /* This branch must be a cbnz, or the last_cti was not fixed up. */
+                if (instr->opcode != OP_cbnz) {
+                    prev = instr_get_prev(instr);
+                    added_size += mangle_indirect_branch_in_trace(
+                        dcontext, trace, instr, target, 0, &delete_after,
+                        instrlist_last(trace));
+                    /* re-check */
+                    instr = prev;
+                    continue;
+                }
+                trace_exit_label = INSTR_CREATE_label(dcontext);
+                ind_target = target;
+                /* Modify the target of the cbnz. */
+                instr_set_target(instr, opnd_create_instr(trace_exit_label));
+                indirect_type = instr_exit_branch_type(instr);
+                /* unset exit type */
+                instr->flags &= ~EXIT_CTI_TYPES;
+                instr_set_our_mangling(instr, true);
+
+                /* Retrieve jump target reg from the xor instruction. */
+                prev = instr_get_prev(instr);
+                if (prev->opcode != OP_eor)
+                    return 0;
+
+                ASSERT(instr_num_srcs(prev) == 4 && opnd_is_reg(instr_get_src(prev, 1)));
+                jump_target_reg = opnd_get_reg(instr_get_src(prev, 1));
+
+                if (!ind_target || jump_target_reg == DR_REG_NULL)
+                    return 0;
+                /* Choose any scratch register except the target reg. */
+                scratch = (jump_target_reg == DR_REG_X0) ? DR_REG_X1 : DR_REG_X0;
+                /* Add the trace exit label. */
+                instrlist_append(trace, trace_exit_label);
+                /* ldr x0, TLS_REG0_SLOT */
+                instrlist_append(
+                    trace,
+                    instr_create_restore_from_tls(dcontext, scratch, TLS_REG0_SLOT));
+                added_size += AARCH64_INSTR_SIZE;
+                /* if x2 alerady contains the jump_target, then there is no need to store
+                 * it away and load value of jump target into it
+                 */
+                if (jump_target_reg != IBL_TARGET_REG) {
+                    /* str x2, TLS_REG2_SLOT */
```

Comment:
Better to use the named constant `IBL_TARGET_REG` for x2.

Reply:
ok, so do you want me to change all of the x2's to `IBL_TARGET_REG` around this function? I do notice that in emit_utils.c there is this comment
```
    *     x2: indirect branch target
     *     TLS_REG0_SLOT: app's x0
     *     TLS_REG1_SLOT: app's x1
     *     TLS_REG2_SLOT: app's x2
```
and to stay consistent the comment above `fixup_indirect_trace_exit` also mentions
```
 * The ibl routine requires:
 *     x2: contains indirect branch target
 *     TLS_REG2_SLOT: contains app's x2
 ```
 and I think it's probably ok to use x2 throughout?

Reply:
Sure, I think it's fine.